src/evict/evict_page.c

```diff
@@ -675,46 +676,45 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
             LF_SET(WT_REC_SCRUB);
     }
 
-    /* Acquire a snapshot if coming through eviction thread route. */
+    /*
+     * Acquire a snapshot if coming through eviction thread route.
+     *
+     * Also, if we have entered eviction through application threads, we will update our snapshot to
+     * evict pages that are not globally visible based on last_running transaction. We will save
+     * some of the running transaction's context, get a new transaction snapshot, perform eviction
+     * and restore the original transaction's context (including snapshot) once we are finished.
+     */
 
     /*
      * TODO: We are deliberately not using a snapshot when checkpoint is active. This will ensure
      * that point-in-time checkpoints have a consistent version of data. Remove this condition once
      * fuzzy transaction ID based checkpoints work is merged.
      */
-    if (FLD_ISSET(evict_flags, WT_REC_EVICTION_THREAD) && !WT_IS_HS(S2BT(session)) &&
-      !conn->txn_global.checkpoint_running) {
-        /*
-         * We rely on the fact that the eviction threads are created with read committed isolation
-         * by default. If this fact doesn't hold anymore in future, we have to force isolation
-         * level.
-         */
-        WT_ASSERT(session, session->txn->isolation == WT_ISO_READ_COMMITTED);
-
-        /*
-         * Eviction threads do not need to pin anything in the cache. We have a exclusive lock for
-         * the page being evicted so we are sure that the page will always be there while it is
-         * being processed. Therefore, we use snapshot API that doesn't publish shared IDs to the
-         * outside world.
-         */
-        __wt_txn_bump_snapshot(session);
-
+    /*
+     * Avoid updating snapshots when application transactions are in final stages of commit or
+     * rollback as they have already released the snapshot. It otherwise makes it harder in later
+     * part of code to detect updates the belonged to last the running transaction from application
+     * thread.
+     */
+    use_snapshot_for_app_thread = !F_ISSET(session, WT_SESSION_INTERNAL) &&
```

Comment:
"use_snapshot_for_app_thread" and "use_snapshot_for_eviction_thread" naming convention signals to me that only in certain cases we use snapshot with application thread and with eviction threads. If my understanding is correct, we also use a snapshot with the application thread unless the metadata is being updated and we always take a snapshot with the eviction thread.

Maybe we can merge the two booleans together into one "acquire_updated_snapshot"

Reply:
I used two boolean to make it easier to understand the conditions required for each scenario and to make the final if condition a bit simpler. If you believe if it's okay to combine the two then I will make the change.

Reply:
I would prefer one boolean. By having two and the complex conditions here, it seems it would be bad if both were true or both were false. Please combine if you can.

Reply:
Each scenario (application and eviction thread) has a different criteria for using snapshot for eviction. I have added an assert to make sure either only one or none of the booleans is set to true (67b7032),